prov/sm2/src/sm2_rma.c

```diff
@@ -0,0 +1,509 @@
+	sm2_generic_format(xfer_entry, ctx->ep->gid, ctx->op, 0, msg->data,
+			   ctx->op_flags, msg->context);
+	cmd_rma = (struct sm2_cmd_sar_rma_msg *) xfer_entry->user_data;
+	memset(cmd_rma, 0, sizeof(*cmd_rma));
+	cmd_rma->sar_hdr.context = ctx;
+
+	payload_used = 0;
+	if (ctx->op == ofi_op_read_req) {
+		cmd_rma->sar_hdr.request_offset = ctx->bytes_requested;
+		ctx->bytes_requested += SM2_RMA_INJECT_SIZE;
+		if (ctx->bytes_requested >= ctx->bytes_total) {
+			ctx->bytes_requested = ctx->bytes_total;
+			cmd_rma->sar_hdr.proto_flags |=
+				FI_SM2_SAR_LAST_MESSAGE_FLAG;
+		}
+		memcpy(cmd_rma->rma_iov, msg->rma_iov,
+		       sizeof(cmd_rma->rma_iov));
+		ctx->msgs_in_flight++;
+
+		return FI_SUCCESS;
+	}
+
+	for (jv = 0; jv < SM2_IOV_LIMIT; jv++) {
+		if (ctx->bytes_sent == ctx->bytes_total)
+			break;
+		bytes_to_send =
+			MIN(msg->msg_iov[0].iov_len, msg->rma_iov[0].len);
+		bytes_to_send =
+			MIN(bytes_to_send, SM2_RMA_INJECT_SIZE - payload_used);
+		if (bytes_to_send == 0)
+			break;
+
+		src = (uint8_t *) msg->msg_iov[0].iov_base;
+		sm2_get_iface_device((void **) msg->desc, &iface, &device);
+		cmd_rma->rma_iov[jv].addr = msg->rma_iov[0].addr;
+		cmd_rma->rma_iov[jv].key = msg->rma_iov[0].key;
+		cmd_rma->rma_iov[jv].len = bytes_to_send;
+		dest = cmd_rma->user_data + payload_used;
+		assert(payload_used < SM2_INJECT_SIZE);
+
+		ret = ofi_copy_from_hmem(iface, device, dest, src,
+					 bytes_to_send);
+		if (ret)
+			return ret;
+		ofi_consume_iov_desc(msg->msg_iov, msg->desc, &msg->iov_count,
+				     bytes_to_send);
+		ofi_consume_rma_iov(msg->rma_iov, &msg->rma_iov_count,
+				    bytes_to_send);
+
+		payload_used += bytes_to_send;
+		ctx->bytes_sent += bytes_to_send;
+	}
+
+	if (ctx->bytes_sent == ctx->bytes_total) {
+		cmd_rma->sar_hdr.proto_flags |= FI_SM2_SAR_LAST_MESSAGE_FLAG;
+	}
+	ctx->msgs_in_flight++;
+
+	return FI_SUCCESS;
+}
+
+/* takes rma command from application, and issues individual rma proto ops. */
+ssize_t sm2_generic_rma(struct sm2_ep *ep, const struct fi_msg_rma *msg,
+			uint32_t op, uint64_t op_flags)
+{
+	ssize_t ret = 0;
+	sm2_gid_t peer_gid;
+	struct sm2_sar_ctx *rma_ctx;
+	struct sm2_xfer_entry *xfer_entry;
+
+	assert(ofi_total_iov_len(msg->msg_iov, msg->iov_count) ==
+	       ofi_total_rma_iov_len(msg->rma_iov, msg->rma_iov_count));
+
+	assert(sizeof(rma_ctx->msg) < SM2_INJECT_SIZE);
+
+	ret = sm2_verify_peer(ep, msg->addr, &peer_gid);
+	if (ret < 0)
+		return -FI_EAGAIN;
+
+	ofi_spin_lock(&ep->tx_lock);
+
+	assert(op == ofi_op_read_req || op == ofi_op_write);
+
+	ret = sm2_alloc_rma_ctx(ep, &rma_ctx);
+	if (ret) {
+		ret = -FI_EAGAIN;
+		goto out;
+	}
+	sm2_fill_rma_ctx(ep, msg, op, op_flags, peer_gid, rma_ctx);
+
+	if (op == ofi_op_read_req && rma_ctx->bytes_total == 0) {
+		uint64_t comp_flags = ofi_tx_cq_flags(ofi_op_read_req);
+		ret = sm2_complete_tx(ep, (void *) msg->context,
+				      ofi_op_read_req,
+				      comp_flags | FI_COMPLETION);
+		ret = FI_SUCCESS;
+		goto out;
+	}
```

Comment:
This 0-byte read looks weird (comment would be nice), also it looks like it leaks a ctx object.

Reply:
I have to remove it now that we claim RMA_EVENT capability.